src/test/fuzz/util/descriptor.cpp

```diff
@@ -84,3 +87,34 @@ bool HasDeepDerivPath(const FuzzBufferType& buff, const int max_depth)
     }
     return false;
 }
+
+bool HasLargeFrag(const FuzzBufferType& buff, const int max_subs)
+{
+    std::stack<int> counts;
+    for (const auto& ch: buff) {
+        if (ch == '(') {
+            counts.push(1);
+        } else if (ch == ',' && !counts.empty()) {
+            if (++counts.top() > max_subs) return true;
+        } else if (ch == ')' && !counts.empty()) {
+            counts.pop();
+        }
+    }
+    return false;
+}
+
+bool HasTooManyWrappers(const FuzzBufferType& buff, const int max_wrappers)
+{
+    auto count{0};
+    // We iterate in reverse order to start counting when we encounter a colon.
+    for (const auto ch: reverse_iterate(buff)) {
+        if (ch == ':') {
+            count++;
+        } else if (ch == ',' || ch == '(' || ch == '{') {
+            count = 0;
+        } else if (count > 0 && ++count > max_wrappers) {
```

Comment:
Is this correct? I'd expect something like:

<details>
<summary>git diff on 178a1da9c4</summary>

```diff
diff --git a/src/test/fuzz/util/descriptor.cpp b/src/test/fuzz/util/descriptor.cpp
index 1ead54f254..434df93994 100644
--- a/src/test/fuzz/util/descriptor.cpp
+++ b/src/test/fuzz/util/descriptor.cpp
@@ -108,12 +108,10 @@ bool HasTooManyWrappers(const FuzzBufferType& buff, const int max_wrappers)
     auto count{0};
     // We iterate in reverse order to start counting when we encounter a colon.
     for (const auto ch: reverse_iterate(buff)) {
-        if (ch == ':') {
-            count++;
+        if (ch == ':' && (++count > max_wrappers)) {
+            return true;
         } else if (ch == ',' || ch == '(' || ch == '{') {
             count = 0;
-        } else if (count > 0 && ++count > max_wrappers) {
-            return true;
         }
     }
     return false;

```
</details>

Reply:
Based on the two inputs that caused the timeouts:
[clusterfuzz-testcase-mocked_descriptor_parse-5224675118481408.bin.not.txt](https://github.com/bitcoin/bitcoin/files/13877884/clusterfuzz-testcase-mocked_descriptor_parse-5224675118481408.bin.not.txt)
    [clusterfuzz-testcase-minimized-descriptor_parse-5888623437217792.bin.txt](https://github.com/bitcoin/bitcoin/files/15238132/clusterfuzz-testcase-minimized-descriptor_parse-5888623437217792.bin.txt)

I think it's supposed to be counting the characters after the colon, not colons themselves. This is also why it iterates in reverse. I don't know quite enough to be sure that all timeout inputs would end up in this format, but based on those two, the function as is seems good.

Reply:
Ooh, right, because each character _is_ a wrapper. That also answers my [question](https://github.com/bitcoin/bitcoin/pull/30197#discussion_r1668972900) about why we iterate in reverse, and why this implementation is correct.

This is probably fairly obvious to people familiar with Miniscript, but feel free to adopt this docstring anyway if it's helpful:

```cpp
    auto count{0};
    // A wrapper is a single character followed by a colon, e.g. `t:`.
    // The colon is dropped between subsequent wrappers; e.g. `dv:` is
    // the `d` wrapper applied to the `v:` wrapper. For that reason, we
    // iterate in reverse and start counting when a colon is found until
    // the start of the fragment is reached.
    for (const auto ch: reverse_iterate(buff)) {
        if (ch == ':') {
            ++count;
        } else if (ch == ',' || ch == '(' || ch == '{') {
            count = 0;  // start of fragment
        } else if (count > 0 && ++count > max_wrappers) {
            return true;
        }
    }
    return false;
```

Reply:
> I don't know quite enough to be sure that all timeout inputs would end up in this format, but based on those two, the function as is seems good.

I can't guarantee it'd avoid all timeouts, but i'm pretty sure it would catch all timeouts which are due to too many nested wrappers.